### ISSUE TITLE
Loading epoll transport from java.library.path fails

### ISSUE THREAD

**Issue by eun-ice** [netty/netty#7069]
### Expected behavior

On restricted linux systems I want to use epoll transport without the JVM being able to write an .so and load it into JVM. For example grsecurity hardened systems do only allow this if the JVM is running as a trusted user (usually root)

### Actual behavior

Putting libnetty-tcnative-linux-x86_64.so in java.library.path works fine, but libnetty_transport_native_epoll.so fails to load.

### Steps to reproduce

Put libnetty_transport_native_epoll.so and libnetty-tcnative-linux-x86_64.so in a path specified in -Djava.library.path (or LD_LIBRARY_PATH env variable).

Start the JVM with a -Dio.netty.native.workdir=/xyz that is mounted with "noexec" with any Netty code using Epoll.

What happens:

libnetty_transport_native_epoll.so fails to load with:

java.lang.UnsatisfiedLinkError: unsupported JNI version 0xFFFFFFFF required by /.../lib/libnetty_transport_native_epoll.so

Initialization of the .so fails, because netty_unix_util_parse_package_prefix in netty_unix_util.c from transport-native-unix-common module looks for "netty-transport-native-epoll" in the file name. However, if the .so is named libnetty-transport-native-epoll.so, it is not loaded at all.

### Minimal yet complete reproducer code (or URL to code)

Any epoll example.

### Netty version

Current trunk.

### JVM version (e.g. `java -version`)

Java(TM) SE Runtime Environment (build 1.8.0_131-b11)

### OS version (e.g. `uname -a`)

Linux 4.9 with grsecurity

### Proposed fixes

I can think of two ways fixing this

1) Modify io.netty.channel.epoll.Native to properly load the lib if it is named libnetty-transport-native-epoll.so

2) Modify netty_unix_util.c to also work if the name of the .so is libnetty_transport_native_epoll.so.

This pull request implements solution 2): https://github.com/netty/netty/pull/7068

**Comment by Scottmitch** [comment]
IIRC @normanmaurer you were going to submit a followup PR to only support the `_` names as a followup to @carl-mastrangelo's PR https://github.com/netty/netty/pull/6884 (see https://github.com/netty/netty/pull/6884#issuecomment-310574798). Once this is done I would imagine this quirk will go away?

**Comment by eun-ice** [comment]
@Scottmitch due to the way the C code currently works it is either `_` or `-`, and currently epoll is loaded from /tmp using a `-` name but from java.library.path it is only loaded using the `_` name.

The PR https://github.com/netty/netty/pull/7068  makes the libraries work with both  (`_` and `-` names), although it fells a littly hacky.

**Comment by Scottmitch** [comment]
> The PR #7068 makes the libraries work with both (_ and - names), although it feels a littly hacky.

IIRC we determined that the `_` variant should would for all use cases. We have control over all the variables here (the library deployments are coupled to the new versions of software, and we don't guarantee older JNI libraries run with newer Java as this is internal API) so if there is no objection I would prefer to keep it simple and support `_`.

**Comment by normanmaurer** [comment]
Yep totally forgot about this... let me do a pr quickly.

That said we will need another tcnative release as well.

Bye
Norman

**Comment by eun-ice** [comment]
Thank you @normanmaurer 

I have no opinion on how the libs schould be named, as long as I can load them from java.library.path.

On Linux the _ names seem to be more typical though.

**Comment by normanmaurer** [comment]
yeah _ sounds good.

**Comment by eun-ice** [comment]
Is there any news on this, should I provide a different pull request?

**Comment by normanmaurer** [comment]
@eun-ice I reached out the the folks that maintain the maven plugin we use and see if we can jus set some config or if we need to add some ant task or such:

https://github.com/fusesource/hawtjni/issues/38

**Comment by eun-ice** [comment]
@normanmaurer thanks. If they do not have a quick solution, maybe you could accept my PR #7068 so at least functionality is restored until a more permanent solution is achieved.

**Comment by eun-ice** [comment]
I think I have to clarify one of my statements.

> , if the .so is named libnetty-transport-native-epoll.so, it is not loaded at all.

That statement seems to imply that libs with `-` can't be loaded at all on grsecurity, but that is not the case. Grsecurity doesn't care what the name of the lib is, it just has to be owned by a trusted user (root) and must not be writable for the user. libnetty-tcnative-linux-x86_64.so loads fine if placed in java.library.path.

The reason why libnetty-transport-native-epoll.so is not loaded seems to be in Native.java. Native.java only tries to load libnetty_transport_native_epoll.so. Linux does load the lib, but initialization fails because it cannot determine the package prefix.

**Comment by normanmaurer** [comment]
@eun-ice so just to make sure I understand correctly you renamed it to use underscores and it failed while when just use it as it is with dashes it just works?

**Comment by eun-ice** [comment]
@normanmaurer  here is the full table:

libnetty-tcnative-linux-x86_64.so works fine with `-`. If renamed to `_` netty tries to load it but loading fails with  `unsupported JNI version 0xFFFFFFFF `

Netty doesn't even try to load libnetty-transport-native-epoll.so it only tries to load libnetty_transport_native_epoll.so . Loading libnetty_transport_native_epoll.so fails with  `unsupported JNI version 0xFFFFFFFF ` to the failure of the package prefix detection.  With my PR #7068  the loading of the lib works in both cases,  with `_` or with `-`

**Comment by eun-ice** [comment]
@normanmaurer I tried to fix Native.java but it already says:

        String []libraryNames = new String[] {
          "netty-transport-native-epoll",
          "netty_transport_native_epoll"
        };
        NativeLibraryLoader.loadFirstAvailable(PlatformDependent.getClassLoader(Native.class), libraryNames);

Still it only tries to load netty_transport_native_epoll and not netty-transport-native-epoll from java.library.path, maybe that's another issue in NativeLibraryLoader.

Because I didn't find a quick way to fix NativeLibraryLoader I simply fixed the .so so it doesn't care about the naming convention. That's what I did in https://github.com/netty/netty/pull/7068

**Comment by eun-ice** [comment]
Fix for NativeLibraryLoader is based on @normanmaurer https://github.com/netty/netty/pull/7091 is here: https://github.com/netty/netty/pull/7097

**Comment by Scottmitch** [comment]
@eun-ice - there are a few PRs which must be pulled in before `netty-transport` native will work. @normanmaurer will have to submit a PR for Netty which makes the loading after PR https://github.com/netty/netty-tcnative/pull/295 in netty-tcnative is pulled in ... stand by until the dust settles related to renaming changes.

**Comment by eun-ice** [comment]
@scottmitch thanks for the update. i have been following the whole story and tried my best to assist, since @normanmaurer s PRs didn't work in my environment. It tried to add a patch but failed at GIT

**Comment by eun-ice** [comment]
Current PR to test this: https://github.com/netty/netty/pull/7102

**Comment by normanmaurer** [comment]
This was fixed by #7102
